### PR TITLE
Fix #3401: Introduced SplashScreenWelcomeMsg PlatformParameter (Compile Time) [BLOCKED on #3387]

### LINKED ISSUE #3401: Introduce a demo PlatformParameter (Compile Time)
Introduce a test PlatformParameter to verify the Compile time Flag architecture.

SplashScreenToast

Target PR date: 6 July 2021
Target completion date: 11 July 2021

### PR DESCRIPTION
**Pull request by Arjupta**
## Explanation

Fixes #3401 
This PR introduces SplashScreenWelcomeMsg PlatformParameter for testing the implementation of Compile Time Platform Parameters. Also this PR is stacked on https://github.com/oppia/oppia-android/pull/3387

## Screenshots
Screenshot 2021-07-07 121030

| Test 1 | Test 2 |
|:-:|:-:|
|Screenshot_1625639975 | Screenshot_1625639982 |

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Comment by Arjupta** [comment]
@vinitamurthi @jcqli I am not ableto resolve these issues 
- I cannot inject PlatformParameterValue\<Boolean\> in SplashTestActivityPresenter as it says Binding is missing (Even after I made PlatformParameterValue to be in Singleton Scope). Thatswhy I went away with a work around of injecting parameter in SplashTestActivity itself
- I am not able to run the Tests as right now compilation fails, the error is of missing Binding for PlatfromParameterSingleton, although I have provided this SingletonImpl in the PlatformParameterModule but still this fails to be detected in SplashTestActivityTest.

**Comment by Arjupta** [comment]
@vinitamurthi @jcqli the tests are working now. PTAL

I want to get your views on ShadowClasses. Can we use ShadowToast in InstrumentationTests , When I tried to so the app didn't compiled saying that unresolved reference for shadows. I also tried to annotate the tests to have @RUNON(ROBOLECTRIC) but still it didn't ran.

File: app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt
Comment (vinitamurthi) [inline app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt]: This can be made private

File: app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt
Comment (vinitamurthi) [inline app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt]: Can you make the log a bit more intuitive? SOmething like 'Failed to load platform parameters, defaults will be used instead.'

File: app/src/main/java/org/oppia/android/app/testing/SplashTestActivity.kt
Comment (vinitamurthi) [inline app/src/main/java/org/oppia/android/app/testing/SplashTestActivity.kt:13]: Add a KDoc

File: app/src/main/java/org/oppia/android/app/testing/SplashTestActivityPresenter.kt
Comment (vinitamurthi) [inline app/src/main/java/org/oppia/android/app/testing/SplashTestActivityPresenter.kt:20]: Add a KDoc

File: app/src/main/java/org/oppia/android/app/testing/SplashTestActivityPresenter.kt
Comment (vinitamurthi) [inline app/src/main/java/org/oppia/android/app/testing/SplashTestActivityPresenter.kt]: This can be made private

File: app/src/sharedTest/java/org/oppia/android/app/administratorcontrols/AdministratorControlsActivityTest.kt
Comment (vinitamurthi) [inline app/src/sharedTest/java/org/oppia/android/app/administratorcontrols/AdministratorControlsActivityTest.kt:83]: Just want to confirm that you checked whether we actually need these changes? I have a strong suspicion that we do, but want to know if you tried running tests without the change. One way to quickly check is to remove it for one of the tests and see if it still passes.

File: app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt
Comment (vinitamurthi) [inline app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt]: Add a KDoc

**Review by vinitamurthi — CHANGES_REQUESTED** [review]

File: app/src/sharedTest/java/org/oppia/android/app/administratorcontrols/AdministratorControlsActivityTest.kt
Comment (Arjupta) [reply app/src/sharedTest/java/org/oppia/android/app/administratorcontrols/AdministratorControlsActivityTest.kt:83]: Yeah I have tried the way you told me to do. I ran the SplashTestActivityTest and then compilation for failed for different tests, so one by one I added this PlatformParameterModule in all the tests. There are no unnecessary changes .

File: app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt
Comment (Arjupta) [reply app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt]: Done

File: app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt
Comment (Arjupta) [reply app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt]: Done

File: app/src/main/java/org/oppia/android/app/testing/SplashTestActivity.kt
Comment (Arjupta) [reply app/src/main/java/org/oppia/android/app/testing/SplashTestActivity.kt:13]: Done

File: app/src/main/java/org/oppia/android/app/testing/SplashTestActivityPresenter.kt
Comment (Arjupta) [reply app/src/main/java/org/oppia/android/app/testing/SplashTestActivityPresenter.kt:20]: Done

File: app/src/main/java/org/oppia/android/app/testing/SplashTestActivityPresenter.kt
Comment (Arjupta) [reply app/src/main/java/org/oppia/android/app/testing/SplashTestActivityPresenter.kt]: Done

File: app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt
Comment (Arjupta) [reply app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt]: Done

File: app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt
Comment (jcqli) [inline app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt]: `getParameterLoadingStatus` sound like a UI process. I'd rename this to something like `fetchPlatformParametersFromDatabase` and then `processParameterLoadingStatus` => `processPlatformParameters` or something to it's clearer that it's data processing.

File: app/src/main/java/org/oppia/android/app/testing/SplashTestActivity.kt
Comment (jcqli) [reply app/src/main/java/org/oppia/android/app/testing/SplashTestActivity.kt:13]: Would it be better to name this `PlatformParametersTestActivity` (and associated presenter & tests)? Since it doesn't matter that it's faking the `SplashActivity`, but it does matter that it's needed to test PlatformParameters.

File: app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt
Comment (jcqli) [inline app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt]:
```
if (windowToken === appToken) {
  // This window isn't contained by any other windows.
```

File: app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt
Comment (jcqli) [inline app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt]: Suggest adding a comment about what this helper is doing.

**Review by jcqli — COMMENTED** [review]
Nice job -- a couple of suggestions mostly for clarity.

File: app/src/main/java/org/oppia/android/app/testing/SplashTestActivity.kt
Comment (vinitamurthi) [reply app/src/main/java/org/oppia/android/app/testing/SplashTestActivity.kt:13]: Thats true actually, both options sound good to me I will defer to what the two of you decide!

Comment (Arjupta) [reply app/src/main/java/org/oppia/android/app/testing/SplashTestActivity.kt:13]: As all the PlatformParameter related task will be triggered by `SplashActivity` only thats why I kept this name for the FakeActivity to represent where is the actual change required. Although I was thinking to rename the `SplashTestActivityTest` to `PlatformParameterIntegrationTest` to represent that we are actually testing PlatformParameter architecture with the help of Fake Activity (SplashTestActivity)

File: app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt
Comment (Arjupta) [reply app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt]: Done

File: app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt
Comment (Arjupta) [reply app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt]: Done

File: app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt
Comment (Arjupta) [reply app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt]: Done

Comment (jcqli) [reply app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt]: This can be written as a regular // comment rather than a KDoc comment. Also, maybe `Returns a [TypeSafeMatcher] which checks for any [Toast] message in the screen.`

Since it's not checking for a specific toast, right?

Comment (jcqli) [reply app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt]: This can be written as a regular // comment rather than a KDoc comment. Also, maybe `Returns a [TypeSafeMatcher] which checks for any [Toast] message in the screen.`

Since it's not checking for a specific toast, right?

File: app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt
Comment (jcqli) [reply app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt]: Please write the comment as a sentence`// Means this window isn't contained by any other windows.`

File: app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt
Comment (Arjupta) [reply app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt]: Yeah, Done

File: app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt
Comment (Arjupta) [reply app/src/sharedTest/java/org/oppia/android/app/testing/SplashTestActivityTest.kt]: Done

File: app/src/main/java/org/oppia/android/app/testing/SplashTestActivity.kt
Comment (jcqli) [reply app/src/main/java/org/oppia/android/app/testing/SplashTestActivity.kt:13]: Hm okay I think it's fine to leave this file name but I do think that the test should be renamed to `PlatformParameterIntegrationTest` or `PlatformParameterTest` since we're not really testing a fake activity, we're testing that we properly fetch parameters. I would also add to the comment that it's a fake SplashActivity since that is what fetches the parameter, something like `A test activity to verify the injection of [PlatformParameter] in the [SplashActivity]. This test activity is used in integration tests for platform parameters.`.

File: app/src/sharedTest/java/org/oppia/android/app/testing/PlatformParameterIntegrationTest.kt
Comment (jcqli) [inline app/src/sharedTest/java/org/oppia/android/app/testing/PlatformParameterIntegrationTest.kt]: nit: [ ] brackets around TypeSafeMatcher

**Review by jcqli — APPROVED** [review]
LGTM, one nit comment

File: app/src/sharedTest/java/org/oppia/android/app/testing/PlatformParameterIntegrationTest.kt
Comment (Arjupta) [reply app/src/sharedTest/java/org/oppia/android/app/testing/PlatformParameterIntegrationTest.kt]: This is not supported as it is not a KDoc. So convertinbg it back KDoc

File: app/src/main/java/org/oppia/android/app/testing/SplashTestActivity.kt
Comment (Arjupta) [reply app/src/main/java/org/oppia/android/app/testing/SplashTestActivity.kt:13]: Done

File: app/src/sharedTest/java/org/oppia/android/app/testing/PlatformParameterIntegrationTest.kt
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/testing/PlatformParameterIntegrationTest.kt:85]: Add espresso test result screenshot

**Review by anandwana001 — CHANGES_REQUESTED** [review]

File: app/src/sharedTest/java/org/oppia/android/app/testing/PlatformParameterIntegrationTest.kt
Comment (Arjupta) [reply app/src/sharedTest/java/org/oppia/android/app/testing/PlatformParameterIntegrationTest.kt:85]: Done. PTAL

**Review by anandwana001 — CHANGES_REQUESTED** [review]
Some of the test cases are failing in checks, PTAL.

**Comment by Arjupta** [comment]
> Some of the test cases are failing in checks, PTAL.

These tests are failing as this PR is not updated with develop. (Blocked on merging its base PR)

File: app/src/main/AndroidManifest.xml
Comment (rt4914) [inline app/src/main/AndroidManifest.xml]: Arrange this in alphabetical order.

**Review by rt4914 — APPROVED** [review]
@ARJUPTA One nit suggested. Thanks.

File: app/src/sharedTest/java/org/oppia/android/app/testing/PlatformParameterIntegrationTest.kt
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/testing/PlatformParameterIntegrationTest.kt:126]: this test is failing for Gradle on robolectric.

**Review by anandwana001 — CHANGES_REQUESTED** [review]

**Comment by anandwana001** [comment]
> > Some of the test cases are failing in checks, PTAL.
> 
> These tests are failing as this PR is not updated with develop. (Blocked on merging its base PR)

ok, so once the base PR is merged here, assign me back.

File: app/src/sharedTest/java/org/oppia/android/app/testing/PlatformParameterIntegrationTest.kt
Comment (Arjupta) [reply app/src/sharedTest/java/org/oppia/android/app/testing/PlatformParameterIntegrationTest.kt:126]: But these tests are running fine on my machine

Comment (anandwana001) [reply app/src/sharedTest/java/org/oppia/android/app/testing/PlatformParameterIntegrationTest.kt:126]:
```
org.oppia.android.app.testing.PlatformParameterIntegrationTest > testSplashTestActivity_updateEmptyDatabase_readDatabaseValues_checkWelcomeMsgIsVisible FAILED
    androidx.test.espresso.NoMatchingRootException: Matcher 'is toast' did not match any of the following roots: [Root{application-window-token=android.view.ViewRootImpl$W@247955c1, window-token=android.view.ViewRootImpl$W@247955c1, has-window-focus=true, layout-params-type=1, layout-params-string={(0,0)(fillxfill) ty=BASE_APPLICATION wanim=0x10302f8
      fl=LAYOUT_IN_SCREEN LAYOUT_INSET_DECOR SPLIT_TOUCH HARDWARE_ACCELERATED DRAWS_SYSTEM_BAR_BACKGROUNDS
      pfl=FORCE_DRAW_STATUS_BAR_BACKGROUND}, decor-view-string=DecorView{id=-1, visibility=VISIBLE, width=960, height=1410, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params={(0,0)(fillxfill) ty=BASE_APPLICATION wanim=0x10302f8
      fl=LAYOUT_IN_SCREEN LAYOUT_INSET_DECOR SPLIT_TOUCH HARDWARE_ACCELERATED DRAWS_SYSTEM_BAR_BACKGROUNDS
      pfl=FORCE_DRAW_STATUS_BAR_BACKGROUND}, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0, child-count=1}}]
        at java.base/java.lang.Thread.getStackTrace(Thread.java:1654)
        at androidx.test.espresso.base.DefaultFailureHandler.getUserFriendlyError(DefaultFailureHandler.java:96)
        at androidx.test.espresso.base.DefaultFailureHandler.handle(DefaultFailureHandler.java:59)
        at androidx.test.espresso.ViewInteraction.waitForAndHandleInteractionResults(ViewInteraction.java:322)
        at androidx.test.espresso.ViewInteraction.check(ViewInteraction.java:306)
        at org.oppia.android.app.testing.PlatformParameterIntegrationTest.testSplashTestActivity_updateEmptyDatabase_readDatabaseValues_checkWelcomeMsgIsVisible(PlatformParameterIntegrationTest.kt:133)
```

1. Could you add a screenshot showing that this robolectric test is passing for you?
2. Could you confirm is the `getToastMatcher` function working fine?
3. COuld you add some reference if any, for the `getToastMatcher()` function?

Comment (Arjupta) [reply app/src/sharedTest/java/org/oppia/android/app/testing/PlatformParameterIntegrationTest.kt:126]: ### 1 and 2 are proven by this screenshot 
Screenshot 2021-07-07 121030

### getToastMatcher( ) 
Answer 2 [Reference-SO](https://stackoverflow.com/questions/28390574/checking-toast-message-in-android-espresso)

File: app/src/main/AndroidManifest.xml
Comment (Arjupta) [reply app/src/main/AndroidManifest.xml]: Done

**Review by vinitamurthi — APPROVED** [review]

**Comment by anandwana001** [comment]
Test is failing locally for me, this is some delay issue, I will try to fix it by today, else we will go through the option for run only on espresso. For integration test. 

By the time, if you come up with something, do let me know. 

For splash test, add espresso result screenshot, if espresso fails, we need to check before merging pr.

**Comment by Arjupta** [comment]
Closing this PR as it cannot be merged directly into develop, all the changes from this PR are cherry-picked to #3441.

**Pull request closed without merging**
